design/nodes-machines-and-hosts.md

```diff
@@ -0,0 +1,213 @@
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives](#alternatives)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-20T17:58-05:00 -->
+
+<!--te-->
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+This document explains the use of Node, Machine, and BareMetalHost
+objects to represent different facets of implementing a self-hosted
+and Kubernetes-native hardware provisioning system. It explains the
+role of each, and why we are using 3 separate resources.
+
+## Motivation
+
+This document details the original design decisions made when the
+MetalKube project was launched, and places the information in context
+as an aid to future contributors.
+
+### Goals
+
+- Explain Node, Machine, and BareMetalHost and how they relate to
+  MetalKube.
+- Encourage input on that design, to refine it.
+
+### Non-Goals
+
+- Discussion of user interface designs related to the resources.
+
+## Proposal
+
+The MetalKube project is building a hardware provisioning system with
+a Kubernetes-native API. It includes a provider implementation that
+fits into the API defined by the cluster-api SIG. It differs from
+other similar implementations by being self-hosted. Where the AWS or
+OpenStack providers run within Kubernetes, they allocate resources
+from a pool *outside* of the Kubernetes cluster. MetalKube will use
+Kubernetes for tracking available resources.
+
+### Implementation Details/Notes/Constraints
+
+#### Data Model
+
+`Node` objects represent a running instance of kubelet. Their status
+fields include basic information about the health of the environment
+in which kubelet is running. A user manipulates the Node to change the
+cluster, such as draining the node for maintenance. The definition of
+Node is owned by the Kubernetes core, and offers few opportunities to
+attach custom data.
+
+`Machine` objects represent a *request* for an instance of kubelet. A
+parallel design is the physical volume claim and physical volume in
+the storage APIs. The Machine CRD is owned by the cluster-api
+SIG. Machine objects are managed by the cluster-api controller using
+an *actuator* which works as a plugin driver for the
+controller. Machine objects include a "provider spec" field to allow
+the actuator to store custom data. The lifecycle of a Machine is based
+on the desired size of the cluster. That is, a Machine object does not
+typically exist in the database when it does not represent a request
+to expand the cluster (see caveats below in the
+[Alternatives](#alternatives-optional) section.
```

Comment:
The cluster-api project publishes some docs here: https://kubernetes-sigs.github.io

This page is particularly relevant: https://kubernetes-sigs.github.io/cluster-api/common_code/machine_controller.html

Reply:
That first link is a 404, but I did link to the second page which works fine.